### DISCUSSION TITLE
Disable SubmitButton on mount

### DISCUSSION THREAD

**Discussion by jrestall** [airjp73/rvf#212] [Q&A]
I understand this goes against recommendations as discussed at https://github.com/airjp73/remix-validated-form/issues/158#issuecomment-1221561650. However we have several forms that only have one or two input fields and the UX team would like the submit button disabled until the forms are valid.

I've tried to implement @airjp73's suggested solution to "call [validate](https://www.remix-validated-form.io/reference/use-form-context#validate) when the form mounts" but I'm struggling with `Error("Validate called before form was initialized.")`. I can get it working on initial SSR by using useHydrated from remix-utils but when doing CSR I can't find a way to determine when the form state is initialized and the page is already hydrated so validate is called before the form is initialized.

```ts
export function SubmitButton(props: ButtonProps) {
  const { form, disabled, ...rest } = props;

  const hydrated = useHydrated();
  const formContext = useFormContext();
  useEffect(() => {
    // Doesn't work with CSR as React is already hydrated so -> "Validate called before form was initialized."
    if(hydrated) formContext.validate();
  }, [hydrated])

  const isValid = useIsValid();
  return <button type="submit" disabled={disabled || !hydrated || !isValid} {...rest} />;
}
```
How can I disable the submit button on mount and until the form is valid in a way that works with both SSR and CSR?

**Comment by airjp73** [comment]
Hi!

I guess calling validate on mount isn't as simple as I thought it was haha. I think you could do this to resolve the error.

```ts
useEffect(() => {
  formContext.validate().catch(() => {});
}, [formContext.validate]);
```

`formContext.validate` will be a different reference after the form is initialized, so the effect will re-run once the form is initialized. I haven't tested it out to 100% confirm.

I'm tempted to dig into the expectations from UX here though. Do they really want to display validation errors for the inputs as soon as the form mounts and bug the user about how they've provided invalid input before they finished typing? I'm wondering if their expectations are more along the lines of "disable the submit button until the user types something in the form", rather than "disable the submit button until the form is valid". If it's the former, you could set some state `onChange` like the [current workaround for tracking `dirty` fields](https://github.com/airjp73/remix-validated-form/issues/104#issuecomment-1292931080).

**Reply by jrestall** [reply]
Hi @airjp73! Appreciate the quick response! I gave your workaround a shot but it doesn't seem to work consistently for me. We have a multistep flow and the form context seems to become uninitialized when moving between forms and then isValid is always false.

UX team are after the later, I suppose so that it is clear when the user has entered the field(s) correctly and can proceed.
